pkg/controller/openapi/openapi_controller.go

```diff
@@ -0,0 +1,399 @@
+	if statusUpdateErr != nil {
+		if reconcileErr != nil {
+			return reconcile.Result{}, fmt.Errorf("Failed to reconcile openapi: %v. Failed to update openapi status: %w", reconcileErr, statusUpdateErr)
+		}
+
+		return reconcile.Result{}, fmt.Errorf("Failed to update openapi status: %w", statusUpdateErr)
+	}
+
+	if statusResult.Requeue {
+		return statusResult, nil
+	}
+
+	if reconcileErr != nil {
+		if helper.IsInvalidSpecError(reconcileErr) {
+			// On Validation error, no need to retry as spec is not valid and needs to be changed
+			reqLogger.Info("ERROR", "spec validation error", reconcileErr)
+			r.EventRecorder().Eventf(openapiCR, corev1.EventTypeWarning, "Invalid OpenAPI Spec", "%v", reconcileErr)
+			return reconcile.Result{}, nil
+		}
+
+		reqLogger.Error(reconcileErr, "Failed to reconcile")
+		r.EventRecorder().Eventf(openapiCR, corev1.EventTypeWarning, "ReconcileError", "%v", reconcileErr)
+		return reconcile.Result{}, reconcileErr
+	}
+
+	return reconcileStatus, nil
+}
+
+func (r *ReconcileOpenAPI) reconcileSpec(openapiCR *capabilitiesv1beta1.OpenAPI) (*OpenAPIStatusReconciler, reconcile.Result, error) {
+	logger := r.Logger().WithValues("openapi", openapiCR.Name)
+
+	err := r.validateSpec(openapiCR)
+	if err != nil {
+		statusReconciler := NewOpenAPIStatusReconciler(r.BaseReconciler, openapiCR, "", err, false)
+		return statusReconciler, reconcile.Result{}, err
+	}
+
+	providerAccount, err := controllerhelper.LookupProviderAccount(r.Client(), openapiCR.Namespace, openapiCR.Spec.ProviderAccountRef, logger)
+	if err != nil {
+		statusReconciler := NewOpenAPIStatusReconciler(r.BaseReconciler, openapiCR, "", err, false)
+		return statusReconciler, reconcile.Result{}, err
+	}
+
+	openapiObj, err := r.readOpenAPI(openapiCR)
+	if err != nil {
+		statusReconciler := NewOpenAPIStatusReconciler(r.BaseReconciler, openapiCR, providerAccount.AdminURLStr, err, false)
+		return statusReconciler, reconcile.Result{}, err
+	}
+
+	err = r.validateOpenAPIAs3scaleProduct(openapiCR, openapiObj)
+	if err != nil {
+		statusReconciler := NewOpenAPIStatusReconciler(r.BaseReconciler, openapiCR, providerAccount.AdminURLStr, err, false)
+		return statusReconciler, reconcile.Result{}, err
+	}
+
+	backendReconciler := NewOpenAPIBackendReconciler(r.BaseReconciler, openapiCR, openapiObj, providerAccount, logger)
+	_, err = backendReconciler.Reconcile()
+	if err != nil {
+		statusReconciler := NewOpenAPIStatusReconciler(r.BaseReconciler, openapiCR, providerAccount.AdminURLStr, err, false)
+		return statusReconciler, reconcile.Result{}, err
+	}
+
+	productReconciler := NewOpenAPIProductReconciler(r.BaseReconciler, openapiCR, openapiObj, providerAccount, logger)
+	_, err = productReconciler.Reconcile()
+	if err != nil {
+		statusReconciler := NewOpenAPIStatusReconciler(r.BaseReconciler, openapiCR, providerAccount.AdminURLStr, err, false)
+		return statusReconciler, reconcile.Result{}, err
+	}
+
+	productSynced, err := r.checkProductSynced(openapiCR)
```

Comment:
Why checkProductSynced is performed and not for backend?

Reply:
The product adds the backend as backend usage. The product will not be in sync until the backend usage is in sync. The product controller makes sure the backend usage's items are valid Backend CRs and are sync'ed.

Reply:
Adding the comment for future reference

Reply:
:+1: